### ISSUE TITLE
stream_snapshot_from_marker is returning None

### ISSUE THREAD

**Issue by dzhulu** [jeog/TOSDataBridge#4]
A few days ago ```stream_snapshot_from_marker``` started returning ```None``` for me, and it was working correctly before. Is anyone else experiencing this issue? The standard ```get()``` works fine, but I'm scared of missing transactions using ```while(runflag): get(...)```.

Here's some test code which generates ```None``` for me via stream_snapshot_from_marker() but the proper transactions via ```get()```. I'm running this via python 3 32-bit but on Windows 10 64-bit.

```
import timer
from timeit import default_timer
buffer = []
tosdb.init(root='C:\TOSDataBridge-master\\bin')
block = tosdb.TOSDB_DataBlock(10000, True)
block.add_items('/ES')
block.add_topics('LASTX', 'BIDX', 'ASKX', 'LAST_SIZE')
time.sleep(3)  # allow block time to load the cache
buffer.append(block.get('/ES', 'BIDX', date_time=True,
                        indx=0))  # update marker
time.sleep(5)  # wait perhaps for transactions to occur
timer = 0
while (timer < 5):  # get transactions for 5 seconds
    #     data = block.get('/ES', 'BIDX', date_time=True) # this works
    data = block.stream_snapshot_from_marker(  # this returns None
        '/ES', 'BIDX', date_time=True, beg=0)
    print(data)  # => prints None
    buffer.append(data)
    timer = default_timer()
tosdb.clean_up()
print(buffer)
# => [('2499.75 G', TOSDB_DateTime(micro=97104, sec=34, min=57,
#        hour=20, day=22, month=9, year=2017)), None, None, None,...]
```

**Comment by dzhulu** [comment]
Hmm I just noticed that the fetched data is totally wrong. I might have a leak somewhere...

**Comment by jeog** [comment]
dzhulu,

What time did this occur? (TOSDB_DateTime is build on your local time and I don't know your timezone.) If you are running this in the (U.S) overnight session (or while the market is closed) it's possible no /ES data came into the stream(s) which would cause the marker calls to return None, especially if you have them in a while loop thats making the calls rapidly over a short time period.

**Comment by dzhulu** [comment]
This occurred today around the time of posting (30 minutes ago when /ES was and still is trading. I restarted my computer and the prints are correct calling stream_snapshot_from_marker in the test code above. However, I still can't get it to work via threaded calls to stream_snapshot_from_marker, which returns only ```None```.

**Comment by jeog** [comment]
So it's working w/ a single thread but not w/ multiple threads?

try using TOSDB_ThreadSafeDataBlock

**Comment by dzhulu** [comment]
Still getting `None` while running `stream_snapshot_from_marker` via `TOSDB_ThreadSafeDataBlock`.

**Comment by jeog** [comment]
And it works fine on the main thread?

**Comment by dzhulu** [comment]
Yes ssfm works in mainthread but is returning None when I call it from separate threads.

Would you recommend I not do multithreading for now? I am simply interested in getting LAST, BID, ASK, LAST_SIZE and coupling them together properly. Perhaps it's overboard using a separate thread to pull each of these four topics?

**Comment by jeog** [comment]
You can- it's a pretty efficient call and doesn't block, plus python uses this thing called the global interpreter lock that keeps all the threads in a single process anyway -  but I'd like to try to figure out what's happening. 

When you use the ssfm call do you have it in a loop that's not waiting/blocking/sleeping/doing heavy processing, so it's getting called rapidly? Because if you just call it in a loop like that you will return alot of None's. Did you happen to change the code recently that caused it to get called a lot more frequently?

**Comment by dzhulu** [comment]
Yes, I have ssfm in an unconditional while loop. I did change the code to run indefinitely and now it doesn't work; whereas, before I ran it for 1-2 seconds and it did work.

I hadn't considered that the `None`s are scrolling past the console so fast that actual transactions went by unnoticed to the eye. D'oh! I'll check this again tomorrow when the market is livelier.

**Comment by jeog** [comment]
Ya, the stream updates pretty fast but not nearly fast enough to keep up with a while loop like that. You're also wasting a lot of processing time, unnecessarily. Try adding a time.sleep(seconds) call inside the loop and see what happens. Something like 1/2 second is probably good enough for most cases.

Maybe at some point I'll add a blocking version that will push data as soon as it gets it but at a low-level it would essentially be the same thing: looping on the call, looking for not None, and sleeping for some arbitrary amount of time before trying again.

**Comment by dzhulu** [comment]
So I put `time.sleep(x)` in the while loop as you recommended and the program works flawlessly. I also used the original `while` loop without sleeping, and the data is also fetched flawlessly, with of course more `None`s. The only difference is I ran the non sleeping `while` code yesterday post lunch and today an hour within market opening, so the data was just not as dense yesterday and I likely had a memory leak since I was testing in Eclipse and the shell. Thanks for your help jeog.
